### PR TITLE
ARROW-8926: [C++] Improve arrow/compute/*.h comments, correct typos and outdated language

### PR DESCRIPTION
**Pull request by wesm**
I also fixed the `Arity` ctor to be explicit which I thought I'd done in #7240 but it is taken care of here.

### CODE REVIEW

File: cpp/src/arrow/compute/kernel.h [hunk 1]
```diff
@@ -345,13 +388,13 @@ class ARROW_EXPORT KernelSignature {
                                                bool is_varargs = false);
 
   /// \brief Return true if the signature if compatible with the list of input
-  /// value descriptors
+  /// value descriptorsl
```
Comment (kiszk) [inline cpp/src/arrow/compute/kernel.h:391]: Is this change necessary?

File: cpp/src/arrow/compute/kernel.h [hunk 2]
```diff
@@ -386,47 +435,75 @@ struct SimdLevel {
...
     INTERSECTION,
 
-    /// Kernel expects a pre-allocated buffer to write the result bitmap into
+    /// Kernel expects a pre-allocated buffer to write the result bitmap
+    /// into. The preallocated memory is not zeroed (except for the last byte),
+    /// so the kernel should ensure to completely populate the bitmap.
     COMPUTED_PREALLOCATE,
 
-    /// Kernel allocates and populates the validity bitmap of the output
+    /// Kernel allocates and set the validity bitmap of the output.
```
Comment (kiszk) [inline cpp/src/arrow/compute/kernel.h:453]: nit: `set` -> `sets`

File: cpp/src/arrow/compute/exec.h [hunk 3]
```diff
@@ -60,29 +60,38 @@ static constexpr int64_t kDefaultExecChunksize = UINT16_MAX;
...
+  /// selecting kernels for execution. Defaults to the library-global function
+  /// registry provided by GetFunctionRegistry.
   FunctionRegistry* func_registry() const { return func_registry_; }
 
-  // \brief Set maximum length unit of work for kernel execution. Larger inputs
-  // will be split into smaller chunks, and, if desired, processed in
-  // parallel. Set to -1 for no limit
+  // \brief Set maximum length unit of work for kernel execution. Larger
+  // contiguous array inputs will be split into smaller chunks, and, if
+  // possible and enable, processed in parallel. The default chunksize is
```
Comment (wesm) [inline cpp/src/arrow/compute/exec.h:80]: enabled

File: cpp/src/arrow/compute/exec.h [hunk 4]
```diff
@@ -146,22 +158,53 @@ class ARROW_EXPORT SelectionVector {
 /// Array and Scalar values and an optional SelectionVector indicating that
 /// there is an unmaterialized filter that either must be materialized, or (if
 /// the kernel supports it) pushed down into the kernel implementation.
+///
+/// ExecBatch is semantically similar to RecordBatch in that in a SQL context
+/// it represents a collection of records, but constant "columns" are
+/// represented by Scalar values rather than having to be converted into arrays
+/// with repeated values.
+///
+/// TOOD: Datum uses arrow/util/variant.h which may be a bit heavier-weight
```
Comment (kiszk) [inline cpp/src/arrow/compute/exec.h:167]: nit: `TOOD` -> `TODO`

File: cpp/src/arrow/compute/kernel.h [hunk 5]
```diff
@@ -324,16 +365,18 @@ class ARROW_EXPORT OutputType {
...
   Resolver resolver_;
 };
 
-/// \brief Holds the input types and output type of the kernel
+/// \brief Holds the input types and output type of the kernel.
 ///
-/// Varargs functions should pass a single input type to be used to validate
-/// the the input types of a function invocation
+/// VarArgs functions should pass a single input type to be used to validate
+/// the the input types of a function invocation.
```
Comment (kiszk) [inline cpp/src/arrow/compute/kernel.h:379]: nit: `the the` -> `the`

File: cpp/src/arrow/compute/kernel.h (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/compute/kernel.h:391]: typo. fixed

File: cpp/src/arrow/compute/kernel.h [hunk 6]
```diff
@@ -386,47 +435,75 @@ struct SimdLevel {
...
 
+/// \brief The strategy to use for propagating or otherwise populating the
+/// validity bitmap of a kernel output.
 struct NullHandling {
   enum type {
     /// Compute the output validity bitmap by intersecting the validity bitmaps
-    /// of the arguments. Kernel does not do anything with the bitmap
+    /// of the arguments. Kernel generally need not touch the bitmap
+    /// thereafter, but a kernel's exec function is permitted to alter the
+    /// bitmap after the null intersection is computed if it needs to.
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernel.h:445]: From this explanation, I find it not fully clear what "intersecting" means. 
(is it where one of the arguments is null, or all or the argument is null, that the output validity indicates null?)

Comment (wesm) [reply cpp/src/arrow/compute/kernel.h:445]: Yes. I'll clarify that it means a bitwise-and operation with all of the respective validity bitmaps

**Comment by wesm** [comment]
Rebasing. I'm going to merge this so avoid further rebase conflicts, but please continue to leave comments and I will address them
